Add IsIPv6UniqueLocal property

annchous opened the pull request:
IsIPv6UniqueLocal boolean property was added to IPAddress. 
In September 2004 was deprecated the definition of the IPv6 site local range. As a result, in October 2005 for use in private IPv6 networks was defined the associated term **unique local addresses**.

Fix [#45148](https://github.com/dotnet/runtime/issues/45148)

Commit:
Add IsIPv6UniqueLocal property

stephentoub commented on src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -485,6 +485,19 @@ public bool IsIPv6Teredo
+            {
+                return IsIPv6 && ((_numbers[0] & 0xFE00) == 0xFC00);
+            }
+        }
```
Thanks for the PR.  To be complete, this also needs to include the new member in the System.Net.Primitives/ref/System.Net.Primitives.cs reference assembly file, and tests.

stephentoub commented on src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -485,6 +485,19 @@ public bool IsIPv6Teredo
             }
         }
 
+        /// <devdoc>
+        ///   <para>
+        ///     Determines if an address is an IPv6 Unique Local address
+        ///   </para>
+        /// </devdoc>
```
```suggestion
        /// <summary>Gets whether the address is an IPv6 Unique Local address.</summary>
```

annchous replied:
Previously created implementations of such properties adhere to the documentation I have provided (for example, `IsIPv6LinkLocal`). Is it exactly necessary to replace this with summary?
If a replacement is necessary, then is it necessary to apply such a style to existing similar properties?

stephentoub replied:
You don't need to update the existing ones: we'll be taking care of that in the coming months. But yes, please use my suggestion instead. Thanks!

annchous replied:
Okay, thank you

danmoseley commented on src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -485,6 +485,19 @@ public bool IsIPv6Teredo
+        {
+            get
+            {
+                return IsIPv6 && ((_numbers[0] & 0xFE00) == 0xFC00);
```
Might need to ! away nullability on _numbers

annchous replied:
```suggestion
                return IsIPv6 && ((_numbers[0]! & 0xFE00) == 0xFC00);
```

Like this, sure? Got it!

Commit:
feat & refactor: after-review fixes

IsIPv6UniqueLocal added in System.Net.Primitives/ref/System.Net.Primitives.cs reference assembly file, tests, away nullability on _numbers and summary instead of devdoc

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=48853) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=48853) before we can accept your contribution.<br/><br/>:x: annchous [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=48853)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=48853) it.</sub>

stephentoub commented on src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -485,6 +485,15 @@ public bool IsIPv6Teredo
+        {
+            get
+            {
+                return IsIPv6 && ((_numbers[0]! & 0xFE00) == 0xFC00);
```
```suggestion
                return IsIPv6 && ((_numbers![0] & 0xFE00) == 0xFC00);
```

annchous replied:
sorry, night coding hurts

Commit:
fix

Commit:
fix: ipv6 address in test

annchous commented:
@stephentoub anything else to fix?)

geoffkizer commented:
LGTM. Thanks!
